src/VisualStudio/Core/Def/Implementation/IntellisenseControls/IntellisenseTextBox.cs

```diff
@@ -0,0 +1,528 @@
+            oleInteropMsg.wParam = msg.wParam;
+            oleInteropMsg.lParam = msg.lParam;
+            oleInteropMsg.pt.x = msg.pt_x;
+            oleInteropMsg.pt.y = msg.pt_y;
+
+            return HandleKeyDown(oleInteropMsg);
+        }
+
+        /// <summary>Converts the given window message for keystroke into a shell command and executes it</summary>
+        /// <param name="message">Message representing the keypress command to handle</param>
+        /// <returns>True if the command is handled (converted and executed properly), false otherwise</returns>
+        /// <remarks>This method is needed when the control is not able to receive VS shell commands directly.
+        /// For instance, it might be hosted in a modal dialog. It generates the proper VS commands 
+        /// and posts it to the internal editor control.</remarks>
+        public bool HandleKeyDown(OLE.Interop.MSG message)
+        {
+            uint editCmdID;
+            Guid editCmdGuid;
+            int VariantSize = 16;
+
+            IVsFilterKeys2 filterKeys = Package.GetGlobalService(typeof(SVsFilterKeys)) as IVsFilterKeys2;
+
+            if (filterKeys != null)
+            {
+                int translated;
+                int firstKeyOfCombo;
+                var pMsg = new OLE.Interop.MSG[1];
+                pMsg[0] = message;
+                ErrorHandler.ThrowOnFailure(filterKeys.TranslateAcceleratorEx(pMsg,
+                    (uint)(__VSTRANSACCELEXFLAGS.VSTAEXF_NoFireCommand | __VSTRANSACCELEXFLAGS.VSTAEXF_UseTextEditorKBScope | __VSTRANSACCELEXFLAGS.VSTAEXF_AllowModalState),
+                    0,
+                    null,
+                    out editCmdGuid,
+                    out editCmdID,
+                    out translated,
+                    out firstKeyOfCombo));
+
+                if (translated == 1)
+                {
+                    IntPtr inArg = IntPtr.Zero;
+                    try
+                    {
+                        // if the command is undo (Ctrl + Z) or redo (Ctrl + Y) then leave it as IntPtr.Zero because of a bug in undomgr.cpp where 
+                        // it does undo or redo only for null, VT_BSTR and VT_EMPTY 
+                        if ((int)message.wParam != Convert.ToInt32('Z') && (int)message.wParam != Convert.ToInt32('Y'))
+                        {
+                            inArg = Marshal.AllocHGlobal(VariantSize);
+                            Marshal.GetNativeVariantForObject(message.wParam, inArg);
+                        }
+
+                        return Exec(ref editCmdGuid, editCmdID, 0, inArg, IntPtr.Zero) == VSConstants.S_OK;
+                    }
+                    finally
+                    {
+                        if (inArg != IntPtr.Zero)
+                            Marshal.FreeHGlobal(inArg);
+                    }
+                }
+            }
+
+            // no translation available for this message
+            return false;
+        }
+    }
+}
```

Comment:
this is an entire file of "oh god" to me.  i can't imagine a good way for the roslyn team to maintain this.

Reply:
yes, we should not maintain this. There are plans to move this to the Editor platform. Then, Roslyn, Debugger and Interactive window can consume it from there. I have added a follow up in another comment.

With this PR, we prototype what we need for Roslyn. Then, we will sync with other parties to extract reusable code.

---
In reply to: [367215852](https://github.com/dotnet/roslyn/pull/38748#discussion_r367215852) [](ancestors = 367215852)